### PR TITLE
chore: Custom Gas Modal Component

### PR DESCRIPTION
**Pull request by blackdevelopa**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**
This PR will introduce a custom gas modal component that should handle the different transaction type for editing gas fees. It's a code improvement PR and so there should be no visible changes to the app UX or UI. This PR will cover the confirm screen for send flow.

Test Case:

**Screenshots/Recordings**
http://recordit.co/1yUAJhqhOy

E2E: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/4b890180-5a54-4361-bc1d-955a03d41f5d

**Issue**

Progresses #???

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

**Comment by bschorchit** [comment]
Hey @gauthierpetetin, did you remove the `needs-dev-review` and added a `in-progress` for this PR by mistake? We currently use the `needs-dev-review` label for internal processes in the team. cc: @hilvmason

**Comment by gauthierpetetin** [comment]
Hi @bschorchit , we setup an automation yesterday to update the label of the Github issue every time the corresponding Zenhub issue is moved to a new column (more detail [here](https://github.com/MetaMask/metamask-zaps/tree/main/zaps/github-labels-automation)). In this case the ticket was moved from "needs-dev-review" to "in-progress" column.

However, I don't understand why it impacts your board, because the automation is only meant to work for the MetaMask Mobile board. I'll investigate and come back to you, sorry for that.

**Comment by gauthierpetetin** [comment]
@bschorchit we've just added an additional filter to make sure the automation only applies to issues with the label "team-mobile-client". It shall not happen again on your board. In case it does, don't hesitate to reach out.

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  isAnimating,
+  onlyGas,
+  validateAmount,
+  updateParent,
+  legacy,
+  legacyGasData,
+  EIP1559GasData,
+  EIP1559GasTxn,
+}: CustomGasModalProps) => {
+  const { colors } = useAppThemeFromContext() || mockTheme;
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:25]: `useAppThemeFromContext() || mockTheme` does not looks good. We should rather change `useAppThemeFromContext` to return mockTheme by default or
`const { colors = mockTheme.colors } = useAppThemeFromContext();`

File: app/components/UI/CustomGasModal/CustomGasModal.test.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            totalMaxHex: '',
+            error: undefined,
+          }}
+          selectedAsset={{ address: '0xABC', symbol: 'ABC', decimals: 0 }}
+        />
+      </Provider>,
+    );
+    expect(wrapper).toMatchSnapshot();
+  });
+});
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.test.tsx:92]: Above is shallow snapshot matching, is not very useful.

File: app/components/Views/SendFlow/Confirm/index.js [hunk 3]
```diff
@@ -1263,9 +1061,7 @@ class Confirm extends PureComponent {
             )}
           </StyledButton>
         </View>
-        {mode === EDIT && this.renderCustomGasModalLegacy()}
         {mode === EDIT_NONCE && this.renderCustomNonceModal()}
-        {mode === EDIT_EIP1559 && this.renderCustomGasModalEIP1559()}
         {this.renderHexDataModal()}
       </SafeAreaView>
     );
```
Comment (jpuri) [inline app/components/Views/SendFlow/Confirm/index.js:1067]: Great cleanup 👏

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+      setEIP1559GasObj(gasObj);
+      updateParent({
+        EIP1559GasTransaction: gasTxn,
+        EIP1559GasObject: gasObj,
+        gasSelectedTemp: selectedGas,
+        gasSelected: selectedGas,
+        closeModal: true,
+      });
+    }
+  };
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:95]: Above function can also be broken down into 2 functions for legacy and 1559 transactions.

**Comment by jpuri** [comment]
@blackdevelopa : while testing with legacy transaction from test DAPP I found that it is showing advance options for legacy transaction:

Uploading Screen Recording 2023-05-01 at 2.28.30 PM.mov…

**Comment by blackdevelopa** [comment]
@jpuri thanks for flagging this. It exists in production and a new issue should be created for it.

> @blackdevelopa : while testing with legacy transaction from test DAPP I found that it is showing advance options for legacy transaction:
> 
>  Screen.Recording.2023-05-01.at.2.28.30.PM.mov

File: app/components/UI/CustomGasModal/CustomGasModal.test.tsx (hunk 2 above)
Comment (blackdevelopa) [reply app/components/UI/CustomGasModal/CustomGasModal.test.tsx:187]: Added some [here](https://github.com/MetaMask/metamask-mobile/pull/6287/commits/90da478ffa41abb258184d2b450bdd2a24e60b0d)

File: app/components/UI/CustomGasModal/CustomGasModal.test.tsx [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+};
+
+const legacyGasData = {
+  legacyGasLimit: '',
+  suggestedGasPrice: '',
+};
+
+describe('CustomGasModal', () => {
+  it('should render correctly', () => {
+    const wrapper = render(
```
Comment (segun) [inline app/components/UI/CustomGasModal/CustomGasModal.test.tsx:103]: Consider using `renderWithProvider` from `app/util/test/renderWithProvider.tsx`

**Review by segun — COMMENTED** [review]
Just one comment.

File: app/components/UI/CustomGasModal/CustomGasModal.test.tsx (hunk 5 above)
Comment (blackdevelopa) [reply app/components/UI/CustomGasModal/CustomGasModal.test.tsx]: Updated. Thank you. https://github.com/MetaMask/metamask-mobile/pull/6287/commits/84c225da3f201dc770b5a03258d498b249b3087e

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+      setEIP1559GasObj(gasObj);
+      updateParent({
+        EIP1559GasTransaction: gasTxn,
+        EIP1559GasObject: gasObj,
+        gasSelectedTemp: selectedGas,
+        gasSelected: selectedGas,
+        closeModal: true,
+      });
+    };
+
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:159]: Nested components declaration is not good idea.

**Review by jpuri — CHANGES_REQUESTED** [review]
Hey, can you plz get rid of nested component declaration. PR looks good otherwise.

**Review by segun — DISMISSED** [review]

**Review by segun — DISMISSED** [review]

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  const getGasAnalyticsParams = () => {
+    try {
+      return {
+        active_currency: { value: selectedAsset.symbol, anonymous: true },
+        gas_estimate_type: gasEstimateType,
+      };
+    } catch (error) {
+      return {};
+    }
+  };
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:58]: Above function can just be constant, its returning static values and really does not needs try/catch block.

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    updateParent({ gasSelected: gas });
+  };
+
+  const onCancelGas = () => {
+    updateParent({
+      stopUpdateGas: false,
+      gasSelectedTemp: selectedGas,
+      closeModal: true,
+    });
+  };
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:71]: Variable names `updateParent`, `stopUpdateGas` are not very intuitive.

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+          legacyGasObject: gasObj,
+          gasSelected: gasSelect,
+          closeModal: true,
+          stopUpdateGas: false,
+          advancedGasInserted: !gasSelect,
+          gasSelectedTemp: gasSelect,
+        });
+      },
+    [transaction, validateAmount, updateParent],
+  );
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:100]: is it possible that we keep `closeModal` in this component itself and not in parent ?
Null can be returned if `closeModal` is true.

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            onCancel={onCancelGas}
+            onSave={onSaveEIP1559GasOption}
+            animateOnChange={animateOnChange}
+            isAnimating={isAnimating}
+            analyticsParams={getGasAnalyticsParams()}
+            view={'SendTo (Confirm)'}
+            selectedGasObject={eip1559GasObject}
+            onlyGas={onlyGas}
+          />
+        )}
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:198]: We can avoid passing here values which are fetched from selector. Those can be obtained in child components.

Comment (blackdevelopa) [reply app/components/UI/CustomGasModal/CustomGasModal.tsx:192]: Yeah. Just checked again and don't think there's any value here that can be gotten from selector that is passed.

File: app/components/UI/CustomGasModal/CustomGasModal.tsx (hunk 8 above)
Comment (blackdevelopa) [reply app/components/UI/CustomGasModal/CustomGasModal.tsx:65]: Yeah. I updated `updateParent` to `updateParentState`. Let me know if this works. I am not sure what a better name for `stopUpdateGas` should be

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+        });
+
+        setEIP1559Txn(gasTxn);
+        setEIP1559GasObj(gasObj);
+        updateParentState({
+          EIP1559GasTransaction: gasTxn,
+          EIP1559GasObject: gasObj,
+          gasSelectedTemp: selectedGas,
+          gasSelected: selectedGas,
+          closeModal: true,
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:118]: @blackdevelopa : parent state params may be different for different components. How will that get handled ?

File: app/components/UI/CustomGasModal/CustomGasModal.tsx [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+          gasSelected: gasSelect,
+          closeModal: true,
+          stopUpdateGas: false,
+          advancedGasInserted: !gasSelect,
+          gasSelectedTemp: gasSelect,
+        });
+      },
+    [transaction, validateAmount, updateParentState],
+  );
+
```
Comment (jpuri) [inline app/components/UI/CustomGasModal/CustomGasModal.tsx:95]: I am bit confused, below looks like callback of callback.

**Review by jpuri — COMMENTED** [review]
Changes look good, just some small questions.

File: app/components/UI/CustomGasModal/CustomGasModal.tsx (hunk 12 above)
Comment (blackdevelopa) [reply app/components/UI/CustomGasModal/CustomGasModal.tsx:72]: Done. [345c539](https://github.com/MetaMask/metamask-mobile/pull/6287/commits/345c539f781c5139988e8e40405e2e137b226071)

File: app/components/UI/CustomGasModal/CustomGasModal.tsx (hunk 11 above)
Comment (blackdevelopa) [reply app/components/UI/CustomGasModal/CustomGasModal.tsx]: Updated. [fd5e14d](https://github.com/MetaMask/metamask-mobile/pull/6287/commits/fd5e14dbc3fe1f289c142bad77619e5a25e39273)

**Review by jpuri — DISMISSED** [review]
Changes look good 👍

**Review by segun — DISMISSED** [review]

**Comment by seaona** [comment]
do you think we could prioritize [this issue](https://github.com/MetaMask/metamask-mobile/issues/5760) soon? It is interfering whenever we test anything related with gas.

cc. @bschorchit @blackdevelopa

**Comment by seaona** [comment]
From QA everything looks good except a small issue I am seeing with Send ETH using a Deeplink.
Basically I can see how the gas does not have a correct value, after re-calculated, making the transaction fail.
Notice: this only happens with Send ETH and not with Send ERC20 token using a deeplink.


https://github.com/MetaMask/metamask-mobile/assets/54408225/68bcf786-1e13-43c0-91a6-dd710a9bbe57



I've tried on `main` and apparently it's working fine. Do you think something from this PR might have affected that function @blackdevelopa ?

1. Select Sepolia
2. Scan this QR code
3. See how gas is set to >0.00000....1
4. Try to Send the tx
5. Tx fails with `Intrinsic gas too low` error

image

**Comment by seaona** [comment]
hi @blackdevelopa ! The issue above seems intermittent and probably not caused by this PR, as we have other reports related to Deeplink gas calculations. For this reason it can be investigated separately.

However I encounter another issue described below.

### Wallet breaks when changing gas priority
Whenever I am sending an ERC20 token, if I edit the gas priority the wallet breaks with the error `Cannot read property 'subst' of undefined`


https://github.com/MetaMask/metamask-mobile/assets/54408225/9f67fad5-4226-4cf3-be39-492aa1e543d8

**Steps**:
1. Import an ERC20 token
2. Click Send
3. Add any amount and recipient
4. Click next
5. Click Edit Gas
6. Change priority Gas
7. Click Save
8. Wallet breaks

**Comment by seaona** [comment]
I can see the issue fixed @blackdevelopa . This looks good from QA. We need x2 re-approvals

**Review by jpuri — DISMISSED** [review]

File: app/components/UI/CustomGasModal/CustomGasModal.test.tsx [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+jest.mock('react-native-keyboard-aware-scroll-view', () => {
+  const KeyboardAwareScrollView = jest.requireActual('react-native').ScrollView;
+  return { KeyboardAwareScrollView };
+});
+
+const gasSelected = 'high';
+const dummyAddress = '0x0';
+
+const initialState = {
```
Comment (cryptotavares) [inline app/components/UI/CustomGasModal/CustomGasModal.test.tsx:25]: There is a [common background initial state](https://github.com/MetaMask/metamask-mobile/blob/main/app/util/test/initial-background-state.json) for unit tests. If possible we should use that one, as it reduces the unit tests maintenance effort.

Comment (blackdevelopa) [reply app/components/UI/CustomGasModal/CustomGasModal.test.tsx]: Great. Updated [here](https://github.com/MetaMask/metamask-mobile/pull/6287/commits/1217799a6b685ffbc659fd9b4f936c03563c39b4)

**Review by jpuri — APPROVED** [review]

**Review by segun — APPROVED** [review]

**Comment by blackdevelopa** [comment]
Merging this PR as it is now QA passed

**Pull request merged**